### PR TITLE
Scan shelf

### PR DESCRIPTION
**Pull request by ederic-oytas**
# Pull Request Template

## Description

Major changes:

- Adds the /scan_shelf endpoint. This endpoint finds all the books in the given image and tries to find those books using Google Books. Currently, the performance is poor, but it "works" as in it gives a result.
- Bugfix: Formerly, the server was sending literal `+`s to Google Books, instead of spaces. This happened because the `requests` library already reformats `+` into `%2B`, so it is read as a literal `+` instead of ` `. Long story short, if two or more fields were specified, `+`s were erroneously added to the Google Books query, messing with it. This fixes that.
- Bugfix: If a user puts a field-like indicator like `inauthor:` inside another field, then this may cause issues. To solve this, we simply add a space before and after the colon: `inauthor : ` by adding a space after substrings like `"inauthor:"`.

Rejected change:
- Rescale images before passing them to EasyOCR. If the image is too big, then we resize it to a smaller size, so EasyOCR takes less time.
  - Reasoning: I've decided against adding rescaling. It seems like it provides too little of a speed boost in trading off accuracy. Also, I cannot figure out how to translate the coordinates properly.
  
Postponed changes:
- Improve performance of /scan_shelf. Currently, the /scan_shelf endpoint performs terribly in terms of accuracy. It does a poor job of finding the correct books in the given image. I believe that the current issue is with sideways text.
  - I've postponed this change until a future PR, so this PR can get merged already.

## Type of Change

- [ ] Bugfix
- [X] New Feature
- [ ] Documentation Update
- [ ] Other (please describe):

## Checklist

- [X] I have read the contributing guidelines.
- [X] I have followed the code style of this project.
- [X] I have added tests to cover my changes.
- [X] I have updated the documentation (if applicable).
- [X] My changes do not generate new warnings.

## Related Issue(s)

#50 
#68 
#69 
#72 
#76

## Additional Notes

i can only blame my shelf

### CODE REVIEW

**Comment by ederic-oytas** [comment]
I solved the formatting issue with Google Books.

**Comment by ederic-oytas** [comment]
I tried to implement image scaling, but I removed it, since I don't think the tradeoff between speed and accuracy is worth it here, after doing some testing. Additionally, I couldn't figure out how to scale coordinates properly.

**Comment by ederic-oytas** [comment]
I've decided to postpone the changes to make the /scan_shelf endpoint more accurate to another PR, so that this PR can be merged already.

File: server/tabby_server/api/books.py [hunk 1]
```diff
@@ -178,6 +182,97 @@ def _get_result_dict(books: list[google_books.Book]) -> dict:
     }
 
 
+@subapp.route("/scan_shelf", methods=["POST"])
+def books_scan_shelf() -> tuple[dict, HTTPStatus]:
+
+    # Load image
```
Comment (LilAttacker) [inline server/tabby_server/api/books.py:188]: Missing docstring

**Review by LilAttacker — APPROVED** [review]
Looks good. Only noticed one of the functions was missing a doc string and the test coverage could be better, but its still good regardless.

Approved.

**Review by caesarx26 — APPROVED** [review]
## Positive Notes:

- The /scan_shelf endpoint is a great addition, even in its draft form. It integrates OCR and Google Books effectively.
- The bug fixes for query formatting and colon handling are simple yet impactful improvements.

## Suggestions for Improvement:

Add tests for:
- /scan_shelf functionality (edge cases like rotated books, empty shelves, or dense shelves).
- Google Books query fixes, especially handling multiple fields and special characters.
Provide baseline performance metrics for /scan_shelf to track future improvements.
Document /scan_shelf API usage and limitations.
- Ensure bounding box scaling is robust by validating it with various edge cases (e.g., books near image edges).

### Chat GPT recommends
[chat gpt link](https://chatgpt.com/share/67479d2e-7030-800a-b3fd-4ca30fa7784f)

**Review by MrSnoodle01 — APPROVED** [review]
I think that adding the scan shelf endpoint is great and should help our app standout. Looks like it returns a similar format as the scan cover does, so implementing this into the frontend should be fairly painless.

**Review by David1000734 — APPROVED** [review]
